src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -88,9 +101,83 @@ public uint DisableSsl(SNIHandle handle)
+            SNITCPHandle tcpHandle = (SNITCPHandle)tdsParserStateObject.Handle;
+
+            SafeDeleteContext securityContext = tdsParserStateObject.sspiClientContextStatus.SecurityContext;
+            ContextFlagsPal contextFlags = tdsParserStateObject.sspiClientContextStatus.ContextFlags;
+            SafeFreeCredentials credentialsHandle = tdsParserStateObject.sspiClientContextStatus.CredentialsHandle;
+
+            SecurityBuffer[] inSecurityBufferArray = null;
+            if (securityContext == null) //first iteration
+            {
+                credentialsHandle = NegotiateStreamPal.AcquireDefaultCredential(Kerberos, false);
+            }
+            else
+            {
+                inSecurityBufferArray = new SecurityBuffer[] { new SecurityBuffer(receivedBuff, SecurityBufferType.SECBUFFER_TOKEN) };
+            }
+
+            int tokenSize = NegotiateStreamPal.QueryMaxTokenSize(Kerberos);
+            SecurityBuffer outSecurityBuffer = new SecurityBuffer(tokenSize, SecurityBufferType.SECBUFFER_TOKEN);
+
+            ContextFlagsPal requestedContextFlags = ContextFlagsPal.Connection
+                | ContextFlagsPal.Confidentiality
+                | ContextFlagsPal.MutualAuth;
+
+            string serverSPN = System.Text.Encoding.UTF8.GetString(serverName);
+
+            SecurityStatusPal statusCode = NegotiateStreamPal.InitializeSecurityContext(
+                       credentialsHandle,
+                       ref securityContext,
+                       serverSPN,
+                       requestedContextFlags,
+                       inSecurityBufferArray,
+                       outSecurityBuffer,
+                       ref contextFlags);
+
+            if (statusCode.ErrorCode == SecurityStatusPalErrorCode.CompleteNeeded ||
+                statusCode.ErrorCode == SecurityStatusPalErrorCode.CompAndContinue)
+            {
+                inSecurityBufferArray = new SecurityBuffer[] { outSecurityBuffer };
+                statusCode = NegotiateStreamPal.CompleteAuthToken(ref securityContext, inSecurityBufferArray);
+            }
+
+            sendBuff = outSecurityBuffer.token;
+
+            tdsParserStateObject.sspiClientContextStatus.SecurityContext = securityContext;
+            tdsParserStateObject.sspiClientContextStatus.ContextFlags = contextFlags;
+            tdsParserStateObject.sspiClientContextStatus.CredentialsHandle = credentialsHandle;
+
+            uint result = SspiClientContextResult.OK;
+            if (IsErrorStatus(statusCode.ErrorCode))
+            {
+                // Kerberos Ticket missing
+                if (statusCode.ErrorCode == SecurityStatusPalErrorCode.InternalError &&
+                    statusCode.Exception is Interop.NetSecurityNative.GssApiException)
+                {
+                    result = SspiClientContextResult.KerberosTicketMissing;
+                }
+                else
+                {
+                    result = SspiClientContextResult.Failed;
```

Comment:
Looks like we lose the statusCode and any Exception it contains.  Seems like that could be useful in helping someone track down why it failed?  Is there any general support for logging already used in SqlClient?  If so that could be used.  Or maybe the exception could be propagated in some way.